labs/b4.md

```diff
@@ -0,0 +1,165 @@
+
+*(For users of Ubuntu: PPAs, or third-party repositories, carry the same risks of running malicious third-party code.)*
+
+If the developers provide a package compatible with your distribution (.deb for Debian-based distros), you can download that package and install it using:
+
+`apt install ./package.deb`
+
+Another way is to find a generic binary package from the developer. This comes in the form of shady shell scripts, or a binary tarball in which you can just extract it and run it, or an appimage, which is a special type of executable that brings its own dependencies.
+
+The last way is to compile your own software from scratch. What does that mean? Free and open-source software usually has their source code somewhere publicly available (GitHub, GitLab, their website). If you fetch their source code, it won’t magically run out of the box. The source code is like the recipe, while the software itself is like food. A package is like a food that is put into a box but we won’t covering the details of making a package yourself (there are tools that do that for you and it varies from platform to platform).
+
+## So how do I compile?
+
+Compiling software on Linux is a mixed bag. Sometimes, all the dependencies are installed on your computer and there is no fiddling around. Sometimes, the dependencies don’t even exist pre-compiled for your distribution so you have to compile those yourself in order to compile what’s at hand. But all of these steps all have one component in common: the makefile.
+
+In most source tarballs, there is usually a makefile that contains a set of directives to compile a project. This is because in a project, there is usually multiple files that need to be compiled by the compiler. And even then, there are multiple compilers and compiler settings that can be used.
+
+Many projects that have to be compiled are usually in C and/or C++. On most Linux distributions, there are usually three compiler “options”. There is the GNU Toolchain which uses GCC, LLVM which uses clang, and Intel’s proprietary toolkit. Both GCC and clang are open-source and free. However, clang was created with a permissive license in response to GCC’s more viral copyleft license. Although clang was made with the intention of being completely compatible with GCC, there are some quirks and most open-source software still target GCC.
+
+To compile most software that you have the dependencies to, run:
+
+`make`
+
+This is will usually choose the correct compiler and compile the whole project. Usually when the compilation is done, the executables are in /bin or /src. 
+
+GCC and clang also have compiler flags that allow certain features to be enabled. Usually the flags that actually matter are optimization flags. Depending on what you want to optimize for, either space or memory or speed, there is a flag for it.
+
+Now, we will make a very simple application in C that prints “Hello Penguin!” named hellopenguin. Invoke:
```

Comment:
I'd say that you are creating a new file here, just to be clear that they don't have to have this file already.